185817277 resize data card

bacalj opened the pull request:
Makes DataCard resize itself to make room for new data as described in [PT#185817277](https://www.pivotaltracker.com/n/projects/2441242/stories/185817277)

It's worth noting that:
- like other tiles, it will not "shrink to fit." 
- The default height is set to something that will fit a typical card with one image and four attributes, but this constant is never used, as we now adjust the height automatically. 

Putting in draft until I resolve an issue with one of the cypress tests
- [x] fix cypress test

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -53,8 +54,23 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer((props) => {
+    if (readOnly) onRequestRowHeight(model.id, height);
+    if (!readOnly){
+      const uiHeight = tileElt?.querySelector(".data-card-container")?.scrollHeight || 0;
+      const heightDiff = height ? height - uiHeight : 0;
+      heightDiff < 30 && onRequestRowHeight(model.id, uiHeight + 60);
```
I'd usually think of a `diff` as change from the old to the new.  So if the tile never shrinks, then I'd think the heightDiff should always be positive. So I'd recommend changing this to:
```suggestion
      const heightDiff = height ? uiHeight - height : 0;
      heightDiff > 30 && onRequestRowHeight(model.id, uiHeight + 60);
```

bacalj replied:
I ended up keeping the calculation intact but changing the names of the variables (and adding a constant).  I did try a version with your calculation but for some reason it was not working, let me know if it makes more sense now or if you still think it should be inverted.

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -53,8 +54,23 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer((props) => {
+  }, [height, model.id, onRequestRowHeight, readOnly, tileElt]);
+
+  useEffect(() => {
+    adjustHeight();
```
Did you try just putting the adjustHeight code directly in this useEffect? It would make the dependency list more clear and doesn't seem like it would make things any more complicated.

The calling of adjustHeight above in the first useEffect seems redundant with this one. They will both be called at the beginning when the component is first rendered. And then this one will be called again when it is re-rendered and the dependencies have changed.

The `currEditAttrId` and `currentEditFacet` dependencies are tricky here. There should be a comment about them. My understanding is that they are listed as dependencies so this will run as the user edits the card. Otherwise it wouldn't run when a new attribute was added.

bacalj replied:
That worked - I think I put it in the loading effect at first for the sake of the left side.

scytacki reviewed (changes requested):
Code looks good, I just requested a few hopefully minor changes.

bacalj commented:
@scytacki I addressed your comments, but I am waiting to re-request review as I have failing cypress test locally (we'll see if it fails in the cloud).  It seems `readOnly` does not apply in the cypress environment and/or my effect timing is out of sync in the cypress environment, because the right side is using the wrong default height (only in cypress), and the drag/merge tests fail because of that.  Will peek at it again later (going offline now for a bit)

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -55,6 +56,15 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer((props) => {
+    if (!tileElt) return;
+    const uiHeight = tileElt?.querySelector(".data-card-container")?.scrollHeight || 0;
+    const spaceLeft = height ? height - uiHeight : 0;
+    if (readOnly) onRequestRowHeight(model.id, Math.max(uiHeight, kExampleDeckHeight));
```
I'm not completely sure, but I think you shouldn't be setting the row height when read only. The row height is saved in the model. So if you are doing this in readOnly it means you are "writting". More concretely if the document is rendered in multiple places (often happens on left side) they will each be trying to update the row height.

lbondaryk replied:
yes but this issue with cards not showing themselves happens in content on the left side as well. It doesn't do anyone any good if you can't see all the stuff in a card - the moth images were getting sliced off halfway even if they were in the first field. So I think you gotta set heights everywhere no? @scytacki

bacalj replied:
Paired briefly with Scott and implemented so that we allow this operation when we are in content.

scytacki replied:
@lbondaryk I'd note that the height will not be adjusted automatically for non-curriculum documents that are being viewed "readOnly". I'd think this will be OK since the height ought to be adjusted when the user original created or edited the document.

bacalj commented:
@scytacki I implemented the change we looked at and indeed the cypress test now passes - I think the failure above is timeout on an unrelated test.

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -1,6 +1,6 @@
 import classNames from "classnames";
 import { observer } from "mobx-react";
-import React, { useEffect, useState } from "react";
+import React, { useCallback, useEffect, useState } from "react";
```
```suggestion
import React, { useEffect, useState } from "react";
```

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -12,18 +12,21 @@ import { SortSelect } from "./components/sort-select";
 import { useConsumerTileLinking } from "../../hooks/use-consumer-tile-linking";
 
 import "./data-card-tile.scss";
+import { getTileEnvironment } from "../../models/tiles/tile-environment";
+import { getDocumentContext } from "../../models/document/document";
```
Both of these should be removed.
```suggestion
```

bacalj replied:
oh thanks, missed those

scytacki reviewed (approved):
This looks good to me. There are couple of lint errors show unused imports that should be fixed. You don't need to re-request my review.

tejal-shah commented:
@bacalj , @scytacki , @lbondaryk , when testing this PR, I find that the data card resizes as expected when I add data/values in the data card against the attributes. But if I just create an empty data card and want to add all my attributes first before filling in the data, it wont resize as I'm adding my attributes. Is this a requirement that should be done as part of this work? Ot should this be in its own story?

tejal-shah commented:
I'm removing the pending QA review label from this PR for the remaining datacard resizing fix in the case I mentioned (as per our discussion in planning today).

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -66,6 +67,14 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer((props) => {
   }, [currEditAttrId, currEditFacet, imageUrlToAdd,
```
@bacalj I'd guess you could just add `content.attributes` to the dependency list here and that would take care of things. I don't think you need the `useRef` and extra useEffect.

bacalj replied:
@scytacki oh, whoops I missed this comment and then pushed what I have below.  I can try what you said above if you think that makes more sense.

bacalj commented:
@scytacki I moved this all out to a hook -  both to get the logic compartmentalized, and because I think there will be more resizing to do once we have the more complex sort view going.

scytacki commented on src/plugins/data-card/data-card-tile.tsx:
```diff
@@ -56,25 +56,17 @@ export const DataCardToolComponent: React.FC<ITileProps> = observer((props) => {
+    documentId,
+    readOnly: readOnly ?? false,
+    onRequestRowHeight,
+    attrCount: attrIdsNames.length,
```
I'd use `content.attributes.length` here instead. `attrIdsNames` will get recreated on each render because it is not a cached view.  Technically what you've got should work the same because the count won't change even if `attrIdsNames` does. But I think it would be more clear if it was using an object that doesn't get recreated each time.

scytacki commented on src/plugins/data-card/use-data-card-tile-height.ts:
```diff
@@ -0,0 +1,44 @@
+    const uiHeight = tileElt.querySelector(".data-card-container")?.scrollHeight || 0;
+    const spaceLeft = height ? height - uiHeight : 0;
+    const requiringHeightChange = attrCtRef.current < attrCount || spaceLeft < kButtonSpace;
+    const loadingInContent = !documentId && readOnly;
```
I'd recommend adding a comment about how the documentId will be undefined when the tile is being rendered in the curriculum tabs.

scytacki commented on src/plugins/data-card/use-data-card-tile-height.ts:
```diff
@@ -0,0 +1,44 @@
+      onRequestRowHeight(modelId, Math.max(uiHeight, kExampleDeckHeight));
+    }
+
+    if (requiringHeightChange) {
```
I'd think you'd want to make this an `else if`. Otherwise even if `loadingInContent` is true, `requiringHeightChange` might be true, and then the height would be updated to include the `kButtonSpace`.

Also if the document is readOnly it seems like `requiringHeightChange` could be true. So I'd guess you'd want to update `requiringHeightChange`'s definition. To be something like:
`const requiringHeightChange = !readOnly && (attrCtRef.current < attrCount || spaceLeft < kButtonSpace);`

scytacki reviewed (changes requested):
I think the separate hook is good. I think I found a few problems with its implementation though.

scytacki reviewed (approved):
This looks good to me.

tejal-shah reviewed (approved):
It seems to work fine. Just FYI, I found a couple of bugs in that when the datacard is linked to a table tile for example, it doesn't resize properly. And since this story said nothing about linking, I created separate bugs for them - [#185986771](https://www.pivotaltracker.com/story/show/185986771) , [#185986662](https://www.pivotaltracker.com/story/show/185986662)
